File: snippets/breadth-first-search/bfs.py
from queue import *

class Graph(object):
    def __init__(self, V):
        self.V = V
        self.adj = [[] for i in range(V)]

    def add_edge(self, src, dest):
        self.adj[src].append(dest)

    def add_bi_edge(self, u, v):
        self.add_edge(u, v)
        self.add_edge(v, u)

    def bfs(self, src):
        visited = [False for i in range(self.V)]
        bfs = Queue(maxsize=self.V)

        visited[src] = True
        bfs.put(src)

        while not bfs.empty():
            v = bfs.get()

            print(v, end=' ')  ###

            for u in self.adj[v]:
                if not visited[u]:
                    visited[u] = True
                    bfs.put(u)

        print()  ###

File: snippets/breadth-first-search/bfs.cpp
#include <bits/stdc++.h>

using namespace std;

class Graph {
public:
    int V;
    vector<list<int>> adj;

    Graph(int V) {
        this->V = V;
        adj.resize(V);
    }

    void AddEdge(int src, int dest) {
        adj[src].push_back(dest);
    }

    void AddBiEdge(int u, int v) {
        AddEdge(u, v);
        AddEdge(v, u);
    }

    void Bfs(int src) {
        vector<bool> visited(V, false);
        queue<int> bfs;

        visited[src] = true;
        bfs.push(src);

        while (!bfs.empty()) {
            int v = bfs.front();
            bfs.pop();

            cout << v << ' ';  ///

            for (const auto& u : adj[v]) {
                if (!visited[u]) {
                    visited[u] = true;
                    bfs.push(u);
                }
            }
        }

        cout << endl;  ///
    }
};
